include/nbl/builtin/hlsl/cpp_compat/promote.hlsl

```diff
@@ -20,76 +23,43 @@ struct Promote
 
 #ifdef __HLSL_VERSION
 
-// TODO: write this scalar-type agnostic!
-template<typename U>
-struct Promote<float32_t1,U>
-{
-    float32_t1 operator()(U v)
-    {
-        return float32_t1(v);
-    }
-};
-
-template<typename U>
-struct Promote<float32_t2, U>
-{
-    float32_t2 operator()(U v)
-    {
-        return float32_t2(v, v);
-    }
-};
-
-template<typename U>
-struct Promote<float32_t3, U>
-{
-    float32_t3 operator()(U v)
-    {
-        return float32_t3(v, v, v);
-    }
-};
-
-template<typename U>
-struct Promote<float32_t4, U>
-{
-    float32_t4 operator()(U v)
-    {
-        return float32_t4(v, v, v, v);
-    }
-};
-
-template<typename U>
-struct Promote<float64_t1,U>
+template<typename Scalar, typename U>
+struct Promote<vector <Scalar, 1>, U>
 {
-    float64_t1 operator()(U v)
+    enable_if_t<is_scalar<Scalar>::value && is_scalar<U>::value, vector <Scalar, 1> > operator()(U v)
     {
-        return float64_t1(v);
+        vector <Scalar, 1> promoted = {Scalar(v)};

```

Comment:
I think you should be able to just write `vector<Scalar,1>(v)` the conversion from U to Scalar is implicit here